File: 468-validate-ip-address/leochang/468.cpp
class Solution {
public:
	string validIPAddress(string IP) {
		if (isIPv4(IP)) {
			return "IPv4";
		} else if (isIPv6(IP)) {
			return "IPv6";
		} else {
			return "Neither";
		}
	}
private:
	bool isIPv4(string IP) {
		if(IP[IP.size()-1] == '.') return false;
		istringstream iss(IP);
		string token;
		int count = 0, v4Count = 0;
		while (getline(iss, token, '.')) {
			bool valid = true;
			for (int i = 0; i < token.size(); i++) {
				if (!isdigit(token[i])) {
					valid = false;
				}
			}
			if (token.size() > 1 && '0' == token[0]) {
				valid = false;
			}
			if (token.size() == 0 || token.size() > 3) {
				valid = false;
			}
			if (valid && stoi(token) < 256) {
				v4Count++;
			}
			count++;
		}
		if (4 == v4Count && 4 == count) {
			return true;
		} else {
			return false;
		}
	}
	bool isIPv6(string IP) {
		if(IP[IP.size()-1] == ':') return false;
		istringstream iss(IP);
		string token;
		int count = 0, v6Count = 0;
		while (getline(iss, token, ':')) {
			if (4 >= token.size() && 0 != token.size()) {
				bool valid = true;
				for (int i = 0; i < token.size(); i++){
					if (string::npos == validIPv6Chars.find(token[i]))
						valid = false;
				}
				if (valid) {
					v6Count++;
				}
			}
			if (0 == count && '0' == token[0]) {
				break;
			}
			count++;
		}
		if (8 == v6Count && 8 == count) {
			return true;
		} else {
			return false;
		}
	}
	const string validIPv6Chars = "0123456789abcdefABCDEF";
};

File: 468-validate-ip-address/ken/468.py
class Solution(object):
    def validIPAddress(self, IP):
        """
        :type IP: str
        :rtype: str
        """
        import re

        def isIPv4(IP):
            pattern = "^(\d+)\.(\d+)\.(\d+)\.(\d+)$"
            matches = re.findall(pattern, IP)

            if len(matches) == 0 or len(matches[0]) != 4:
                return False

            for match in matches[0]:
                if match.startswith('0') and len(match) > 1:
                    return False
                if int(match) > 255:
                    return False

            return True

        def isIPv6(IP):
            pattern = "^([\dabcdefABCDEF]+):([\dabcdefABCDEF]+):([\dabcdefABCDEF]+):([\dabcdefABCDEF]+):([\dabcdefABCDEF]+):([\dabcdefABCDEF]+):([\dabcdefABCDEF]+):([\dabcdefABCDEF]+)$"
            matches = re.findall(pattern, IP)

            if len(matches) == 0 or len(matches[0]) != 8:
                return False

            for match in matches[0]:
                if len(match) > 4:
                    return False

            return True

        if isIPv4(IP):
            return "IPv4"

        if isIPv6(IP):
            return "IPv6"

        return "Neither"
